lib_static/plugins/acts_as_journalized/lib/cause_of_change/base.rb

```diff
@@ -0,0 +1,56 @@
+#-- copyright
+# OpenProject is an open source project management software.
+# Copyright (C) 2012-2023 the OpenProject GmbH
+#
+# This program is free software; you can redistribute it and/or
+# modify it under the terms of the GNU General Public License version 3.
+#
+# OpenProject is a fork of ChiliProject, which is a fork of Redmine. The copyright follows:
+# Copyright (C) 2006-2013 Jean-Philippe Lang
+# Copyright (C) 2010-2013 the ChiliProject Team
+#
+# This program is free software; you can redistribute it and/or
+# modify it under the terms of the GNU General Public License
+# as published by the Free Software Foundation; either version 2
+# of the License, or (at your option) any later version.
+#
+# This program is distributed in the hope that it will be useful,
+# but WITHOUT ANY WARRANTY; without even the implied warranty of
+# MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
+# GNU General Public License for more details.
+#
+# You should have received a copy of the GNU General Public License
+# along with this program; if not, write to the Free Software
+# Foundation, Inc., 51 Franklin Street, Fifth Floor, Boston, MA  02110-1301, USA.
+#
+# See COPYRIGHT and LICENSE files for more details.
+#++
+
+module CauseOfChange
+  class Base
+    attr_reader :type, :additional_attributes
+
+    def initialize(type, additional_attributes = {})
+      @type = type
+      @additional_attributes = additional_attributes
+    end
+
+    def to_h
+      { "type" => type }.merge(additional_attributes).deep_stringify_keys
+    end
```

Comment:
Defining the implicit conversion method `to_hash`  instead of the explicit `to_h` yields the benefit of automatic type conversion based on the callee method.
```suggestion
    def to_hash
      { "type" => type }.merge(additional_attributes).deep_stringify_keys
    end
    
    alias_method :to_h, :to_hash # alias it just for the convenience if we ever want to use the explicit conversion
```

For example there is non need to explicitly call `to_h` in the `ActiveSupport::JSON.encode(cause&.to_h || {})` call. It can be written as `ActiveSupport::JSON.encode(cause || {})`, and the receiver method will implicitly take care of the conversion if the argument object responds to `to_hash`.

This implicit conversion also happens, when comparing the cause object against a hash. The comparator will implicitly calls `to_hash` on the non-hash object and will use returned value if the comparison succeeds. In case there is no `to_hash` method on the object, it will raise an error that it cannot compare the 2 objects.